src/sentry/api/paginator.py

```diff
@@ -220,3 +224,101 @@ def get_result(self, limit=100, cursor=None):
             next=next_cursor,
             prev=prev_cursor,
         )
+
+
+def reverse_bisect_left(a, x, lo=0, hi=None):
+    """\
+    Similar to ``bisect.bisect_left``, but expects the data in the array ``a``
+    to be provided in descending order, rather than the ascending order assumed
+    by ``bisect_left``.
+
+    The returned index ``i`` partitions the array ``a`` into two halves so that:
+
+    - left side: ``all(val > x for val in a[lo:i])``
+    - right side: ``all(val <= x for val in a[i:hi])``
+    """
+    if lo < 0:
+        raise ValueError('lo must be non-negative')
+
+    if hi is None:
+        hi = len(a)
+
+    while lo < hi:
+        mid = (lo + hi) // 2
+        if a[mid] > x:
+            lo = mid + 1
+        else:
+            hi = mid
+
+    return lo
+
+
+class SequencePaginator(object):
+    def __init__(self, data, reverse=False):
+        self.scores, self.values = map(
+            list,
+            zip(*sorted(data, reverse=reverse)),
+        ) if data else ([], [])
+        self.reverse = reverse
+        self.search = functools.partial(
+            bisect.bisect_left if not reverse else reverse_bisect_left,
+            self.scores,
+        )
+
+    def get_result(self, limit, cursor=None):
+        if cursor is None:
+            cursor_score = None
+            cursor_offset = 0
+            cursor_is_prev = False
+        else:
+            cursor_score = cursor.value
+            cursor_offset = cursor.offset
+            cursor_is_prev = cursor.is_prev
+
+        assert cursor_offset > -1
+
+        if cursor_score is None:
+            position = 0 if not cursor_is_prev else len(self.scores)
+        else:
+            position = self.search(cursor_score)
+
+        position = position + cursor_offset
+
+        if not cursor_is_prev:
+            lo = max(position, 0)
+            hi = min(lo + limit, len(self.scores))
+        else:
+            # TODO: It might make sense to ensure that this hi value is at
+            # least the length of the page + 1 if we want to ensure we return a
+            # full page of results when paginating backwards while data is
+            # being mutated.
```

Comment:
Design decision needed here: I could see an argument being made either way to support either case. I didn't check to see what the other paginator does, but precedent might be a valid decision criteria here?